sycl/source/detail/queue_impl.cpp

```diff
@@ -26,9 +26,59 @@
 namespace sycl {
 inline namespace _V1 {
 namespace detail {
-
 std::atomic<unsigned long long> queue_impl::MNextAvailableQueueID = 0;
 
+static std::vector<sycl::detail::pi::PiEvent>
+getPIEvents(const std::vector<sycl::event> &DepEvents,
+            const EventImplPtr &ExtraDepEvent) {
+  std::vector<sycl::detail::pi::PiEvent> RetPiEvents;
+  auto AddEvent = [&RetPiEvents](const EventImplPtr &EventImpl) {
+    if (EventImpl->getHandleRef() == nullptr)
+      return;
+    RetPiEvents.push_back(EventImpl->getHandleRef());
+  };
+  if (ExtraDepEvent)
+    AddEvent(ExtraDepEvent);
+  for_each(DepEvents.begin(), DepEvents.end(),
+           [&RetPiEvents, &AddEvent](const sycl::event &Event) {
+             auto EventImpl = detail::getSyclObjImpl(Event);
+             return AddEvent(EventImpl);
+           });
+  return RetPiEvents;
+}
+
+static bool isEventsReady(const std::vector<sycl::event> &DepEvents,
+                          const EventImplPtr &ExtraDepEventPtr,
+                          ContextImplPtr Context) {
+  auto CheckEvent = [&Context](const EventImplPtr &SyclEventImplPtr) {
+    // throwaway events created with empty constructor will not have a
+    // context (which is set lazily) calling getContextImpl() would set that
+    // context, which we wish to avoid as it is expensive.
+    if (!SyclEventImplPtr->isContextInitialized() &&
+        !SyclEventImplPtr->is_host()) {
+      return true;
+    }
+    if (SyclEventImplPtr->is_host()) {
+      return SyclEventImplPtr->isCompleted();
+    }
+    // Cross-context dependencies can't be passed directly.
```

Comment:
They way the comment is worded right now I'd expect a deadlock here... Can we be more elaborate here? My best guess would be that this entire routine is an "optimization" path and we don't need to provide a precise "false" answer, only a precise "true" one. If so, that should into the documentation of the function.

Reply:
I think the actual problem here was the misleading function name. Changed it to `canBypassScheduler` and made small updates to the comments.